feat(evm): add database with separate storage for block data

DracoLi opened the pull request:
## Why this should be merged

Adds a `ethdb.Database` implementation that moves EVM block data out of the shared KV store into dedicated height-indexed databases optimized for block storage, reducing KV store size and compaction cost.

- Headers, bodies, and receipts are stored in `heightindexdb`; all other data remains in the underlying KV store.
- A decrease in total LevelDB/PebbleDB size and total compaction time is expected. Detailed metrics will be tracked in #4523.

## How this works

The `blockdb.Database` wraps `ethdb.Database` and routes block data by key prefix, while leaving non-block data in the KV store.

- Header/body/receipt keys for blocks with height ≥ the configured minimum are written to height-indexed DBs; below that threshold they use the KV store.
- At most one block per height is stored. Writing at an occupied height overwrites the previous value and is not an intended use case. Deletes of block data are no-ops.
- If a block isn’t yet in the height-indexed DBs and migration isn’t complete, reads fall back to the KV store.
- Block-related keys bypass the KV batch and write directly to the height-indexed DBs; non-block keys are batched on the underlying KV store.
- Deferred init: With `allowDeferredInit`, initialization can wait until the minimum block height is known (e.g., when state sync is enabled). Database min height is persisted and cannot be changed without recreating the block databases.

### Migration

Separated to https://github.com/ava-labs/avalanchego/pull/4771

## How this was tested

- Unit test for routing and migration behavior.
- Two Mainnet nodes (with state sync enabled and disabled) with `blockdb` enabled running for >4 weeks.

## Need to be documented in RELEASES.md?

No

DracoLi committed:
feat(evm): add database with separate block storage

ARR4N commented on vms/evm/database/blockdb/database.go:
```diff
@@ -0,0 +1,528 @@
+	heightindexdb "github.com/ava-labs/avalanchego/x/blockdb"
+)
+
+var (
```
Although it's idiomatic in other languages, it's not a requirement in Go to declare all globals at the top like this. The primary drawback is that they lack context relative to where they're used so a reader has to jump back and forth between declaration and usage sites.

Please move the declarations (perhaps still in groups, but also feel free to break them up if warranted) closer to where they're used.

ARR4N commented on vms/evm/database/blockdb/database.go:
```diff
@@ -0,0 +1,528 @@
+const (
+	hashDataElements = 2
+	blockNumberSize  = 8
+	blockHashSize    = 32
```
(fyi) To better explain my rationale for removing the other constants, I think these are useful because they couple multiple usage sites that all MUST match.

DracoLi commented on vms/evm/database/blockdb/database.go:
```diff
@@ -0,0 +1,528 @@
+	ethdb.Database
+
+	// Databases
+	stateDB    database.Database
```
agreed. how about metaDB or configDB. this is used for storing database meta so that might be more appropriate.

DracoLi commented on vms/evm/database/blockdb/database.go:
```diff
@@ -0,0 +1,528 @@
+	_ ethdb.Database = (*Database)(nil)
+	_ ethdb.Batch    = (*batch)(nil)
+
+	migratorDBPrefix = []byte("migrator")
```
this is the prefix on the `stateDB`, which is different than the `evmDB`, which contains the evm data. The `evmDB` currently has the `ethedb` prefix and we should be passing a different prefixed db when creating the `blockdb.Database` (See a draft usage of this database [here](https://github.com/ava-labs/avalanchego/pull/4627/files#diff-9eec6fc986677ba415f585b88b7487d622715323e58d9097cd52d77138c94ee4R132-R133))

DracoLi commented on vms/evm/database/blockdb/database.go:
```diff
@@ -0,0 +1,528 @@
+			return nil, false, err
+		}
+		return db, true, nil
+	}
```
This is an interesting way to rewrite this that I have never thought about. thx for this suggestion, I agree this is easier to read.

DracoLi commented on vms/evm/database/blockdb/database_test.go:
```diff
@@ -0,0 +1,526 @@
+
+func TestMain(m *testing.M) {
+	customtypes.Register()
+	params.RegisterExtras()
```
This is needed by the `core.GenerateChainWithGenesis` call in `createBlocksToAddr` in `helpers_test.go`.
Without them I get a nil pointer error on `coreth/params.GetExtra` and `coreth/plugin/evm/customtypes.GetHeaderExtra`

JonathanOppenheimer replied:
I am resolving this.

DracoLi committed:
fix: updates from feedback

DracoLi committed:
remove migrator logic

DracoLi committed:
fix: remove usage of coreth for creating blocks

DracoLi committed:
fix: make noop delete a debug instead

evm is expected to reject side-chain blocks so having this as warn is not accurate

DracoLi committed:
Merge branch 'master' into dl/evm-blockdb

DracoLi committed:
fix: license header lint

JonathanOppenheimer commented on vms/evm/database/blockdb/database.go:
```diff
@@ -0,0 +1,425 @@
+// the databases.
+//
+// Returns an error if already initialized.
+func (db *Database) InitBlockDBs(minHeight uint64) error {
```
there's no possibility of concurrent calls this function right?

JonathanOppenheimer reviewed (commented):
A few comments/questions

DracoLi committed:
chore: add godocs to exported methods

DracoLi commented on vms/evm/database/blockdb/database.go:
```diff
@@ -0,0 +1,425 @@
+// Copyright (C) 2019, Ava Labs, Inc. All rights reserved.
```
https://github.com/ava-labs/avalanchego/pull/4485/commits/93c1534cc2304e21033435e50dd68220e714d7f6

DracoLi committed:
chore: remove evm prefix from prefix consts

JonathanOppenheimer commented on vms/evm/database/blockdb/database.go:
```diff
@@ -0,0 +1,436 @@
+	}
+
+	var hdb database.HeightIndex
+	switch key[0] {
```
I think this needs some more validation checks. `parseKey` switches on `key[0]` but you don't actually check `len(key) > 0` so calls like `Get(nil)` will panic once height-index DBs are initialized.

The wrapper should handle empty keys by returning false and delegating I think unless you have a better idea (we certainly don't want to panic!)

JonathanOppenheimer commented on vms/evm/database/blockdb/database.go:
```diff
@@ -0,0 +1,436 @@
+
+// Has retrieves if a key is present in the key-value data store.
+// Block data is routed to the height-indexed databases.
+func (db *Database) Has(key []byte) (bool, error) {
```
You certainly know _vastly_ more about databases than I do, but for block keys, you call `readBlock` which does the full `Get` and then decode, which seems expensive. Is there a better way to do this? (no idea, maybe not lol)

JonathanOppenheimer commented on vms/evm/database/blockdb/database.go:
```diff
@@ -0,0 +1,436 @@
+	headerPrefix    = 'h'
+	blockBodyPrefix = 'b'
+	receiptsPrefix  = 'r'
+)
```
I was talking to Stephen about this today (unexported stuff from geth). I'm still not much of a fan of doing this, idk. We've had PRs like this: https://github.com/ava-labs/libevm/pull/262

JonathanOppenheimer replied:
I know this has already been discussed a bunch on this PR.

JonathanOppenheimer commented on vms/evm/database/blockdb/database.go:
```diff
@@ -0,0 +1,425 @@
+	errInvalidEncodedLength = errors.New("invalid encoded length")
+)
+
+var _ ethdb.Database = (*Database)(nil)
```
resolving this

JonathanOppenheimer commented on vms/evm/database/blockdb/database.go:
```diff
@@ -0,0 +1,436 @@
+// Block data is written directly to the height-indexed databases, bypassing the batch.
+func (b *batch) Put(key []byte, value []byte) error {
+	if p, ok := b.db.parseKey(key); ok {
+		return p.writeHashAndData(value)
```
1. Canonical currently says height `H` -> hash `A`.
2. A new candidate hash `B` at height `H` is staged in batch.
3. `batch.Put` writes block data for `B` immediately (overwriting height `H` slot).
4. Batch is discarded/fails, so canonical mapping remains `H`

So now, the DB says canonical hash is `A`, but payload at height `H` is `B`; reads via canonical path can fail (`ErrNotFound` on hash mismatch)

alarso16 commented on vms/evm/database/blockdb/database.go:
```diff
@@ -0,0 +1,436 @@
+
+// parseKey parses a block key into a parsedBlockKey.
+// It returns false if no block databases for the key prefix exist.
+func (db *Database) parseKey(key []byte) (*parsedBlockKey, bool) {
```
I think you should just return whether a nil key if it's not writable
```suggestion
func (db *Database) parseKey(key []byte) *dbKey {
```

alarso16 reviewed (commented):
It's somewhat difficult to reason about. At minimum, you should improve the documentation and naming a little (e.g., what is a "min height"? we also at least have a genesis. Do you mean first non-genesis block committed here?). It was also unclear to me until I read through most of this that this wrapped an ethdb.Database with blocksdb, since it's not described anywhere and appears to only contain blockdb logic (and is in blockdb package). Documenting this better would be immensely helpful

alarso16 commented on vms/evm/database/blockdb/database.go:
```diff
@@ -0,0 +1,436 @@
+
+var blockPrefixes = []byte{blockBodyPrefix, headerPrefix, receiptsPrefix}
+
+func parseBlockKey(key []byte) (num uint64, hash common.Hash, ok bool) {
```
Is this also copied from libevm?

DracoLi commented:
closing - I will revisit the integration after SAE
